prettify-ts-plugin/src/plugin.ts

```diff
@@ -0,0 +1,130 @@
+    // `ts.LanguageServer` and `tsserver.LanguageService` are essentially the same type because they're both created from the same thing in TypeScript's definition files.
+    // However an enum like `SyntaxKind` doesn't obey normal structural equivalence rules and so trying to assign `ts.LanguageServer` to `tsserver.LanguageService` will fail.
+    // The way this line gets around this is essentially by generating the type `Partial<tsserver.LanguageService> & ts.LanguageService` instead.
+    // Later by returning, it checks if it's assignable to `tsserver.LanguageService`. If the types were to fall out of sync this wouldn't work.
+    const proxy = Object.assign<Partial<tsserver.LanguageService>, ts.LanguageService>({}, plugin.languageService)
+
+    proxy.getQuickInfoAtPosition = plugin.getQuickInfoAtPosition.bind(plugin)
+
+    return proxy
+  }
+
+  getQuickInfoAtPosition (fileName: string, position: number): ts.QuickInfo | undefined {
+    const tsserverQuickInfo = this.languageService.getQuickInfoAtPosition.bind(
+      this.languageService
+    )
+
+    if (!this.config.enableHover) {
+      return tsserverQuickInfo(fileName, position)
+    }
+
+    const { ts, program, typeChecker } = this
+
+    const sourceFile = getValidSourceFile(program, fileName)
+    const node = ts.getTouchingPropertyName(sourceFile, position)
+    if (node === sourceFile) return undefined
+
+    const nodeForQuickInfo = getNodeForQuickInfo(ts, node)
+    const symbol = getSymbolAtLocationForQuickInfo(this.ts, nodeForQuickInfo, typeChecker)
+
+    if (symbol === undefined || typeChecker.isUnknownSymbol(symbol)) {
+      if (shouldGetType(this.ts, sourceFile, nodeForQuickInfo, position)) {
+        const type = typeChecker.getTypeAtLocation(nodeForQuickInfo)
+
+        return {
+          kind: ts.ScriptElementKind.unknown,
+          kindModifiers: ts.ScriptElementKindModifier.none,
+          textSpan: ts.createTextSpanFromNode(nodeForQuickInfo, sourceFile),
+          displayParts: this.runWithCancellationToken(typeChecker => {
+            return ts.typeToDisplayParts(typeChecker, type, ts.getContainerNode(nodeForQuickInfo))
+          }),
+          documentation: type.symbol?.getDocumentationComment(typeChecker),
+          tags: type.symbol?.getJsDocTags(typeChecker)
+        }
+      } else {
+        return undefined
+      }
+    }
+
+    const { symbolKind, displayParts, documentation, tags } = this.runWithCancellationToken(typeChecker => {
+      const enclosingDeclaration = ts.getContainerNode(nodeForQuickInfo)
+
+      return ts.SymbolDisplay.getSymbolDisplayPartsDocumentationAndSymbolKind(typeChecker, symbol, sourceFile, enclosingDeclaration, nodeForQuickInfo)
+    })
+
+    return {
+      kind: symbolKind,
+      kindModifiers: ts.SymbolDisplay.getSymbolModifiers(typeChecker, symbol),
+      textSpan: ts.createTextSpanFromNode(nodeForQuickInfo, sourceFile),
+      displayParts,
+      documentation,
+      tags
+    }
```

Comment:
To actually be useful, the code here has to be changed to actually include the prettification logic. I started the change but I backed out of it here because it ended up being complex to do in a way that satisfied me. I think the `ts-morph` solution could in theory be ported quite straightforwardly but I simply haven't used `ts-morph` before and I don't know if it'll actually increase performance.

It seems like it reparses the source file a lot which reads to me like it's contributing to the slowness. However I wasn't quite sure what the best way to expand the type without it was, hence this draft.